### ISSUE TITLE
Codegen from openAPI spec?

### ISSUE THREAD

**Issue by zeskeertwee** [zeskeertwee/destiny-rs#2]
Generating code from the openAPI spec (found [here](https://github.com/Bungie-net/api)) could be very interesting, as it would allow for updating the library quickly after the spec is updated.

**Comment by zperk13** [comment]
Started working on it... well at least struct generation for non-manifest types. I have a general idea of how it would work. I've been coding for over 2 hours and it's almost 2AM and I need to sleep. Do you want to make a destiny-rs-codegen repo or something?

**Comment by zeskeertwee** [comment]
I think we could also just add a crate to this repo, after putting destiny-rs in it's own folder. Or is that a terrible idea?

**Comment by zperk13** [comment]
> I think we could also just add a crate to this repo, after putting destiny-rs in it's own folder. Or is that a terrible idea?

I dont understand what you mean, could you rephrase?

**Comment by zeskeertwee** [comment]
Basically, having a cargo workspace in this repository with 2 folders, destiny_rs and destiny_rs_codegen.

Does that make any sense?

**Comment by zperk13** [comment]
Yes, though I'm not sure how that would react with crates.io. Also that makes it sound like destiny_rs_codegen would be a library, whereas I'm imaging it being a binary. What I have in mind also won't be able to fully automate the proccess, just help out a bit.

**Comment by zeskeertwee** [comment]
Destiny-rs would still be the only crate published (unless the codegen is used in a build script ofc), so that wouldn't change.

**Comment by zperk13** [comment]
I've never published anything on crates.io before so you'd know better than me. Do what you think is best.

**Comment by zperk13** [comment]
Just finished writing the main part of the codegen. It generated a 5,436 line rust file. It's not done, but it's a pretty good start.
Things that need to be done:

- Make the id_to_struct_name() function actually do what it says
- Add the proper formatting to the "mapped to" comments
- Generate Enumerations
- Generate MobileManifestEntities
- Rename properties named "type" since that's a reserved keyword


edit: Oh and I think some code needs to be added to handle the Entities that end in []. I didn't look too much into those. There's only 3 of them. I had it print out some numbers:

Objects generated: 533
Enumerations not generated: 133
Manifests not generated: 51
Object[] things skipped: 3

**Comment by zeskeertwee** [comment]
That's awesome, and I'd gladly help to implement some of the missing things!

**Comment by zperk13** [comment]
Where should I put the code?

**Comment by zeskeertwee** [comment]
You can just create a new folder in this repo and put the project there

**Comment by zperk13** [comment]
Alright. I added Enumeration generation, made the bungie_docs_link() method work, and made the camel_case_to_snake_case() function work. Made PR #10. You may want to change where it writes the output file. I also didn't touch the workspace stuff. Oh and the enumeration thing brings the number of lines in the generated file to 8,227

**Comment by zeskeertwee** [comment]
I'm damn impressed you managed to write it that quickly, and that it works this well already!
I'm curious as to why you chose to parse the HTML documentation and not the openAPI spec?

**Comment by zperk13** [comment]
Thanks! I'm not sure what you mean by the "openAPI spec".

**Comment by zperk13** [comment]
Oh I found it... that might've been helpful lol

**Comment by zeskeertwee** [comment]
I haven't looked a lot at the code yet but from what i've seen it shouldn't be too hard to use that instead of the HTML?
(you know better than me though!)

**Comment by zperk13** [comment]
In some ways, I'd rather use regex to parse HTML than deal with the very unspecific error messages from serde_json

**Comment by zeskeertwee** [comment]
Oh that's odd, in my experience serde_json always told me the exact line the error was at

**Comment by zperk13** [comment]
Maybe they are and I'm just not good at understanding it. If you'd prefer to use the json file over the html, that's fine, but I'll need some help.

**Comment by zeskeertwee** [comment]
that's fine, i'll take a look at it!
